build_tools/benchmarks/collect_compilation_statistics.py

```diff
@@ -0,0 +1,194 @@
+
+The benchmark suites need to be built with ninja and enable the CMake option
+IREE_ENABLE_COMPILATION_BENCHMARKS.
+"""
+
+import argparse
+from dataclasses import dataclass
+import json
+import os
+import re
+import zipfile
+from typing import Dict, Optional
+from pathlib import PurePath
+
+from common.benchmark_definition import CompilationInfo, CompilationResults, CompilationStatistics, ModuleComponentSizes, get_git_commit_hash
+from common.benchmark_suite import BENCHMARK_SUITE_REL_PATH, BenchmarkSuite
+
+BENCHMARK_FLAGFILE = "flagfile"
+MODULE_DIR = "vmfb"
+MODULE_FILE_EXTENSION = ".vmfb"
+NINJA_LOG_HEADER = "ninja log v5"
+NINJA_BUILD_LOG = ".ninja_log"
+COMPILATION_STATS_MODULE_SUFFIX = "compile-stats"
+
+VM_COMPONENT_NAME = "module.fb"
+CONST_COMPONENT_NAME = "_const.bin"
+DISPATCH_COMPONENT_PATTERNS = [
+    r".+_embedded_elf_.+\.so", r".+_vulkan_spirv_fb\.fb",
+    r".+_vmvx_bytecode_fb\.bin"
+]
+
+
+def match_module_cmake_target(module_path: str) -> Optional[str]:
+  # Get the last 4 parts of module path. They are expected to be:
+  # benchmark_suites/<category>/vmfb/<module filename>.vmfb
+  path_parts = PurePath(module_path).parts[-4:]
+  if len(path_parts) < 4:
+    return None
+  if path_parts[0] != BENCHMARK_SUITE_REL_PATH:
+    return None
+  if path_parts[2] != MODULE_DIR:
+    return None
+  if os.path.splitext(path_parts[3])[1] != MODULE_FILE_EXTENSION:
+    return None
+  return os.path.join(*path_parts)
+
+
+def parse_compilation_time_from_ninja_log(log_path: str) -> Dict[str, int]:
+  """Retrieve the compilation time from the Ninja build log."""
+
+  target_build_time_map = {}
+  with open(log_path, "r") as f:
+    header = f.readline()
+    if NINJA_LOG_HEADER not in header:
+      raise NotImplementedError(f"Unsupported ninja log version: {header}")
+
+    for line in f:
+      start_time, end_time, _, target, _ = line.strip().split("\t")
+      cmake_target = match_module_cmake_target(target)
+      if cmake_target is None:
+        continue
+
+      start_time = int(start_time)
+      end_time = int(end_time)
+      target_build_time_map[cmake_target] = end_time - start_time
+
+  return target_build_time_map
+
+
+def get_module_component_info(module_path: str) -> ModuleComponentSizes:
+  module_zipinfo = zipfile.ZipFile(module_path)
+  size_map = dict(
+      (info.filename, info.file_size) for info in module_zipinfo.infolist())
+  vm_component_size = size_map[VM_COMPONENT_NAME]
+  const_component_size = size_map[CONST_COMPONENT_NAME]
+  identified_names = {VM_COMPONENT_NAME, CONST_COMPONENT_NAME}
+  total_dispatch_component_size = 0
+  for filename, size in size_map.items():
+    for pattern in DISPATCH_COMPONENT_PATTERNS:
+      if re.match(pattern, filename):
+        total_dispatch_component_size += size
+        identified_names.add(filename)
+        break
+
+  if identified_names != set(size_map.keys()):
+    raise AssertionError(
+        f"Unrecognized components in the module: {size_map.keys()}.")
```

Comment:
This shouldn't be an assertion error, since it comes from parsing external input.